doc/md/declarative: add doc for declarative migrations with single schemas (HCL) (#1451)
ariga/atlas: doc/md/declarative/declarative-hcl.mdx

---
id: declarative-hcl-singleschema
title: Getting started with HCL declarative migrations for a MySQL schema
slug: /declarative/declarative-hcl
---

import InstallationInstructions from '../components/_installation_instructions.mdx'

### Introduction

Atlas makes it easy to manage your database using the declarative migrations workflow. Declarative migrations focus on describing the desired state of the database and allow the migration engine to plan and execute the necessary actions to reach that state.

In this quickstart guide, we will go through the three essential steps for using Atlas to manage your database schema: inspecting your database, authoring the desired changes, and applying those changes.

### Installing Atlas

<InstallationInstructions />

### Warming up

> To demonstrate the examples in the article, let's create a docker container with an `example` database and insert a table in it.

Reviewer: ```suggestion
To demonstrate the examples in the article, let's run a docker container with an `example` database and create a table in it.
```

> Use the following command to run a MySQL Docker container

Reviewer: Where is the command?

Reviewer: Take it from here:
https://atlasgo.io/getting-started#start-a-local-database-container

```console
docker exec -it atlas-demo mysql -ppass --database=example
```

> Once you're connected, you can add a table in it by running the following command:

Reviewer: ```suggestion
Once you're connected, you can create a table in it by running the following command:
```

```sql
CREATE TABLE customers (
    id INT PRIMARY KEY AUTO_INCREMENT,
    first_name VARCHAR(255) NOT NULL,
    last_name VARCHAR(255) NOT NULL
);
```

## Step 1: Inspecting the database

Before planning any changes to your database, it can be useful to know what your database schema looks like. To do this, you can use the `atlas schema inspect` command to inspect your database. This command generates a file that represents your database schema, which can be used as a starting point for making changes.

### Usage:
```
atlas schema inspect [flags]
```

### Example:

> The following command will inspect the current state of the schema `example` and store it in a file named schema.hcl

Reviewer: ```suggestion
The following command will inspect the current state of the schema `example` and store it in a file named `schema.hcl`
```

```console
atlas schema inspect -u "mysql://root:pass@localhost:3306/example" > schema.hcl
```

```hcl title="schema.hcl"
schema "example" {}

table "customers" {
  schema = schema.example
  column "id" {
    null           = false
    type           = int
    auto_increment = true
  }
  column "first_name" {
    null = false
    type = varchar(255)
  }
  column "last_name" {
    null = false
    type = varchar(255)
  }
  primary_key {
    columns = [column.id]
  }
}
```
:::info
To learn more about inspecting schema with Atlas in variety of use cases, visit the documentation [here](https://atlasgo.io/declarative/inspect)
:::

## Step 2: Planning a change

Once you have inspected your database, the next step is to plan a change. This is done by modifying the HCL file in a way that it defines the desired state of your schema. The file should include the changes you want to make to your database, such as adding tables, columns, or indexes.

### Example:

Let’s add the following table definition to the HCL schema file that we previously created with the `atlas schema inspect` command:

> ```hcl

Reviewer: Show the complete file and use `diff` to show the exact changes.

table "orders" {
  schema = schema.example
  column "id" {
    null           = false
    type           = int
    auto_increment = true
  }
  column "customer_id" {
    null = false
    type = int
  }
  column "order_date" {
    null = false
    type = date
  }
  primary_key {
    columns = [column.id]
  }
  foreign_key "customer_id_fk" {
    columns     = [column.customer_id]
    ref_columns = [table.customers.column.id]
  }
}
```

## Step 3: Applying the change

> The final step is to apply the change to your database. To do this, simply run the `atlas schema apply` command with the required flags. This command will compare the current state of your database to the desired state defined in your HCL file and make the necessary changes.

Reviewer: ```suggestion
The final step is to apply the change to your database. To do this, simply run the `atlas schema apply` command. This command will compare the current state of your database to the desired state defined in your HCL file and make the necessary changes.
```

---
id: declarative-hcl-singleschema

> title: A Quickstart Guide to Declarative Migrations with a Single Schema

Reviewer: @rotemtam, propose a title please

Reviewer: _Getting started with HCL declarative migrations for MySQL schema_
_Managing MySQL schema with Declarative migrations using HCL_

Reviewer 2: > _Getting started with HCL declarative migrations with a MySQL schema_ 

Sounds good to me!

Reviewer 3: sounds good

slug: /declarative/declarative-hcl
---

import InstallationInstructions from '../components/_installation_instructions.mdx'

### Introduction

Atlas makes it easy to manage your database using the declarative migrations workflow. Declarative migrations focus on describing the desired state of the database and allow the migration engine to plan and execute the necessary actions to reach that state.

In this quickstart guide, we will go through the three essential steps for using Atlas to manage your database schema: inspecting your database, authoring the desired changes, and applying those changes.

### Installing Atlas

<InstallationInstructions />

### Warming up
To demonstrate the examples in the article, let's create a docker container with an `example` database and insert a table in it.
Use the following command to run a MySQL Docker container

```console
docker run --name atlas-demo -e MYSQL_ROOT_PASSWORD=pass -e MYSQL_DATABASE=example -p 3306:3306 -d mysql:latest
```

Once the container is running, you can connect to the MySQL server using the statement below:

```console

> docker exec -it atlas-demo mysql -ppass
> ```
> 
> Once you're connected, you can switch to the `example` database and add a table in it by running the following command:
> 
> ```sql
> USE example;

Reviewer: ```suggestion
docker exec -it atlas-demo mysql -ppass --database=example
```

Once you're connected, you can add an example table in it by running the following command:

```sql
```

Reviewer 3: 👍

CREATE TABLE customers (
    id INT PRIMARY KEY AUTO_INCREMENT,
    first_name VARCHAR(255) NOT NULL,
    last_name VARCHAR(255) NOT NULL
);
```

## Step 1: Inspecting the database

> Before making any changes to your database, it's important to know what your database schema looks like. To do this, you can use the `atlas schema inspect` command to inspect your database. This command generates a file that represents your database schema, which can be used as a starting point for making changes.

Reviewer 3: ```suggestion
Before planning any changes to your database, it can be useful to know what your database schema looks like. To do this, you can use the `atlas schema inspect` command to inspect your database. This command generates a file that represents your database schema, which can be used as a starting point for making changes.
```

Reviewer 3: also link to schema inspect full doc

### Usage:
```
atlas schema inspect [flags]
```

### Example:

The following command will inspect the current state of the schema `example` and store it in a file named schema.hcl

```console
atlas schema inspect -u "mysql://root:pass@localhost:3306/example" > schema.hcl
```

```hcl title="schema.hcl"
schema "example" {}

table "customers" {
  schema = schema.example
  column "id" {
    null           = false
    type           = int
    auto_increment = true
  }
  column "first_name" {
    null = false
    type = varchar(255)
  }
  column "last_name" {
    null = false
    type = varchar(255)
  }
  primary_key {
    columns = [column.id]
  }
}
```

> ## Step 2: Authoring the desired change

Reviewer 3: ```suggestion
## Step 2: Planning a change
```

> Once you have inspected your database, the next step is to author the desired change. This is done by modifying the HCL file that defines the present state of your schema. The file should include the changes you want to make to your database, such as adding tables, columns, or indexes.

Reviewer 3: ```suggestion
Once you have inspected your database, the next step is to plan a change. This is done by modifying the HCL file that defines the desired state of your schema. The file should include the changes you want to make to your database, such as adding tables, columns, or indexes.
```

Reviewer 3: In this example we use HCL but you can also use SQL (link)

### Example:

> Let’s add the following table definition to the HCL schema file that we previously created by `atlas schema inspect` command:

Reviewer 2: ```suggestion
Let’s add the following table definition to the HCL schema file that we previously created with the `atlas schema inspect` command:
```

```hcl
table "orders" {
  schema = schema.example
  column "id" {
    null           = false
    type           = int
    auto_increment = true
  }
  column "customer_id" {
    null = false
    type = int
  }
  column "order_date" {
    null = false
    type = date
  }
  primary_key {
    columns = [column.id]
  }

>   foreign_key "orders_ibfk_1" {

Reviewer: give it a more meaningful name.
```suggestion
  foreign_key "customer_id" {
```

    columns     = [column.customer_id]
    ref_columns = [table.customers.column.id]
  }
}
```

> ## Step 3: Applying the changes

Reviewer 3: ```suggestion
## Step 3: Applying the change
```

> The final step is to apply the changes to your database. To do this, simply run the `atlas schema apply` command with required flags. This command will compare the current state of your database to the desired state defined in your HCL or SQL file and make the necessary changes.

Reviewer 3: ```suggestion
The final step is to apply the change to your database. To do this, simply run the `atlas schema apply` command with the required flags. This command will compare the current state of your database to the desired state defined in your HCL file and make the necessary changes.
```

### Usage:

```console
atlas schema apply [flags]
```

### Example:

Let’s apply the changes we created in Step 2 using the following command:

```console
atlas schema apply \
  --url "mysql://root:pass@localhost:3306/example" \
  --to "file://schema.hcl"
```

> Atlas compares  the current state of the database with the state described in the schema file, and shows you any necessary changes to bring the database in line with the schema definition.

Reviewer: ```suggestion
Atlas compares the current state of the database with the state described in the schema file, and shows you any necessary changes to bring the database in line with the schema definition.
```

Reviewer: > Atlas compares  the current state of the database with the state described in the schema file

Maybe omit it from above if it's written here?

```console
-- Planned Changes:
-- Create "orders" table
CREATE TABLE `example`.`orders` (`id` int NOT NULL AUTO_INCREMENT, `customer_id` int NOT NULL, `order_date` date NOT NULL, PRIMARY KEY (`id`), INDEX `orders_ibfk_1` (`customer_id`), CONSTRAINT `orders_ibfk_1` FOREIGN KEY (`customer_id`) REFERENCES `example`.`customers` (`id`) ON UPDATE NO ACTION ON DELETE NO ACTION) CHARSET utf8mb4 COLLATE utf8mb4_0900_ai_ci;
Use the arrow keys to navigate: ↓ ↑ → ← 
? Are you sure?: 
  ▸ Apply
    Abort
```

> Selecting `Apply` will apply the changes to the local environment.

Reviewer 2: ```suggestion
Selecting `Apply` will apply the changes to the database.
```

> You can inspect the latest state of the database into the schema.hcl file again and verify that the changes have been applied by re-running the `atlas schema inspect` command:

Reviewer 2: ```suggestion
You can inspect the latest state of the database again and verify that the changes have been applied by re-running the `atlas schema inspect` command:
```

```console
atlas schema inspect -u "mysql://root:pass@localhost:3306/example"
```
```hcl
table "customers" {
  schema = schema.example
  column "id" {
    null           = false
    type           = int
    auto_increment = true
  }
  column "first_name" {
    null = false
    type = varchar(255)
  }
  column "last_name" {
    null = false
    type = varchar(255)
  }
  primary_key {
    columns = [column.id]
  }
}
table "orders" {
  schema = schema.example
  column "id" {
    null           = false
    type           = int
    auto_increment = true
  }
  column "customer_id" {
    null = false
    type = int
  }
  column "order_date" {
    null = false
    type = date
  }
  primary_key {
    columns = [column.id]
  }
  foreign_key "orders_ibfk_1" {
    columns     = [column.customer_id]
    ref_columns = [table.customers.column.id]
    on_update   = NO_ACTION
    on_delete   = NO_ACTION
  }
  index "orders_ibfk_1" {
    columns = [column.customer_id]
  }
}
schema "example" {
  charset = "utf8mb4"
  collate = "utf8mb4_0900_ai_ci"
}
```

> We can confirm that the table has been effectively inserted into the `example` schema.

Reviewer 3: ```suggestion
Great! Our changes have been successfully applied to the `example` schema.
```

> :::info

Reviewer 2: I think this info block can be a part of the conclusion instead of an info block.

By following the steps above, you can inspect the current state of the database, make changes to the schema definition file and apply those changes to the database with confidence.
:::

### Conclusion

> The declarative migration is a flexible way to manage your database schema. It allows you to define your desired state in a simple, human-readable language, and then automatically apply those changes to your database. This can save you time and effort, and also reduce the risk of errors or inconsistencies in your database schema.

Reviewer 3: link to conceptual articles for deeper reads: declarative vs versioned, SQL vs hcl. we want to let the user know that atlas can work in other ways as well

---
id: declarative-hcl-singleschema

> title: Getting started with HCL declarative migrations with a MySQL schema

Reviewer: ```suggestion
title: Getting started with HCL declarative migrations for a MySQL schema
```
